hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotCache.java

```diff
@@ -0,0 +1,375 @@
+   */
+  public ReferenceCounted<IOmMetadataReader, SnapshotCache> get(String key,
+      boolean skipActiveCheck) throws IOException {
+    // Atomic operation to initialize the OmSnapshot instance (once) if the key
+    // does not exist.
+    ReferenceCounted<IOmMetadataReader, SnapshotCache> rcOmSnapshot =
+        dbMap.computeIfAbsent(key, k -> {
+          LOG.info("Loading snapshot. Table key: {}", k);
+          try {
+            return new ReferenceCounted<>(cacheLoader.load(k), false, this);
+          } catch (OMException omEx) {
+            // Return null if the snapshot is no longer active
+            if (!omEx.getResult().equals(FILE_NOT_FOUND)) {
+              throw new IllegalStateException(omEx);
+            }
+          } catch (IOException ioEx) {
+            // Failed to load snapshot DB
+            throw new IllegalStateException(ioEx);
+          } catch (Exception ex) {
+            // Unexpected and unknown exception thrown from CacheLoader#load
+            throw new IllegalStateException(ex);
+          }
+          // Do not put the value in the map on exception
+          return null;
+        });
+
+    if (rcOmSnapshot == null) {
+      // The only exception that would fall through the loader logic above
+      // is OMException with FILE_NOT_FOUND.
+      throw new OMException("Snapshot table key '" + key + "' not found, "
+          + "or the snapshot is no longer active",
+          OMException.ResultCodes.FILE_NOT_FOUND);
+    }
+
+    // If the snapshot is already loaded in cache, the check inside the loader
+    // above is ignored. But we would still want to reject all get()s except
+    // when called from SDT (and some) if the snapshot is not active anymore.
+    if (!skipActiveCheck &&
+        !omSnapshotManager.isSnapshotStatus(key, SNAPSHOT_ACTIVE)) {
+      throw new OMException("Unable to load snapshot. " +
+          "Snapshot with table key '" + key + "' is no longer active",
+          FILE_NOT_FOUND);
+    }
+
+    // Increment the reference count on the instance.
+    rcOmSnapshot.incrementRefCount();
+
+    synchronized (pendingEvictionList) {
+      // Remove instance from clean up list when it exists.
+      pendingEvictionList.remove(rcOmSnapshot);
+      // Check if any entries can be cleaned up.
+      // At this point, cache size might temporarily exceed cacheSizeLimit
+      // even if there are entries that can be evicted, which is fine since it
+      // is a soft limit.
+      cleanup();
+    }
+
+    return rcOmSnapshot;
+  }
+
+  /**
+   * Release the reference count on the OmSnapshot instance.
+   * @param key snapshot table key
+   */
+  public void release(String key) {
+    ReferenceCounted<IOmMetadataReader, SnapshotCache> rcOmSnapshot =
+        dbMap.get(key);
+    Preconditions.checkNotNull(rcOmSnapshot,
+        "Key '" + key + "' does not exist in cache");
+
+    if (rcOmSnapshot.decrementRefCount() == 0L) {
+      synchronized (pendingEvictionList) {
```

Comment:
If another thread does a get() between lines 228 and 229, it will increment the count and we will still remove the snapshot below.

Reply:
I believe you mean "it will **decrement** the count"?

Should be fixed by the [change](https://github.com/apache/ozone/pull/4567#discussion_r1239066138) mentioned above because ref count is now increased as soon as it gets the RC instance from `dbMap`.

Reply:
Done in 5da12ac5c0593c2a12fce367163a972b5d3f6de0

Reply:
> I believe you mean "it will decrement the count"?

No I meant the other thread doing the get() will increment the count.

Let me try to clarify.  Suppose thread 1 is doing a release(). It runs the "if 0" test above, finds the count is 0, and starts to enter the sychronized block.  But the lock is taken so it waits.  Meanwhile thread 2 takes control and does a get().

Now the reference count is 1, and control returns to thread 1.  That thread now enters the synchronized block, adds the snapshot to the eviction list, and does a cleanup() which could evict the snapshot from the cache.  But the refcount is still 1, and thread 2 is still going to try to use the snapshot without reloading it into the cache.

Isn't that possible?

And moving the "if 0" test into the synchronized block won't help, because increment() isn't protected by that lock.  So the increment() could happen anytime during the synchronized block.

Reply:
I see. That could indeed be an issue.

> And moving the "if 0" test into the synchronized block won't help, because increment() isn't protected by that lock. So the increment() could happen anytime during the synchronized block.

Putting `decrementRefCount()` inside a `synchronized (dbMap)` should work now that 5da12ac5c0593c2a12fce367163a972b5d3f6de0 is in. Because:

1. `incrementRefCount()` is now only done in the `compute()` lambda.
2. `dbMap.compute()` effectively runs the lambda inside a `synchronized (dbMap)`.